ast/term.go

```diff
@@ -1090,29 +1090,35 @@ type QueryIterator func(map[Var]Value, Value) error
 
 // ArrayTerm creates a new Term with an Array value.
 func ArrayTerm(a ...*Term) *Term {
-	return &Term{Value: &Array{elems: a, hash: 0, ground: termSliceIsGround(a)}}
+	return NewTerm(NewArray(a...))
 }
 
 // NewArray creates an Array with the terms provided. The array will
 // use the provided term slice.
 func NewArray(a ...*Term) *Array {
-	return &Array{elems: a, hash: 0, ground: termSliceIsGround(a)}
+	hs := make([]int, len(a))
+	for i, e := range a {
+		hs[i] = e.Value.Hash()
+	}
+	return &Array{elems: a, hashs: hs, ground: termSliceIsGround(a)}
 }
 
 // Array represents an array as defined by the language. Arrays are similar to the
 // same types as defined by JSON with the exception that they can contain Vars
 // and References.
 type Array struct {
 	elems  []*Term
-	hash   int
+	hashs  []int
```

Comment:
This means the `Hash()` function has to recompute the hash code for the array... I wonder if we'll run into some surprising performance issues if we do this... how could we cache the hash and avoid this?

Reply:
I'd suspect this is an issue only if the average array size is fairly high. Then again, if we are about to increase the memory usage anyway, perhaps allocating one more int is not that of a big deal.

Reply:
Went with adding another int field for the computed array hash.